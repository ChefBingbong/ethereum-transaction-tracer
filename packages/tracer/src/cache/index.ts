export * from './abiCache'
export * from './abiSources'
