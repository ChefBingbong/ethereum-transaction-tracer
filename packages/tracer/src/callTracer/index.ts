export * from './tracer'
export * from './types'
