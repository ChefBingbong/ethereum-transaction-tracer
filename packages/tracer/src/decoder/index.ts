export * from './callDecoder'
export * from './decoder'
export * from './precompileDecoder'
