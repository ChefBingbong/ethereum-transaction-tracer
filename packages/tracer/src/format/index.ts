export * from './formatCall'
export * from './formatDelegateCall'
export * from './formatPreCompile'
export * from './formatReturn'
export * from './theme'
