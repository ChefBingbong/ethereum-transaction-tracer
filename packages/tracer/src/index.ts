export * from './cache'
export * from './callTracer'
export * from './decoder'
export * from './format'
export * from './token-slot-override'
