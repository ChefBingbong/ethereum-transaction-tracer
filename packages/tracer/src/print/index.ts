export * from './call'
export * from './gas'
