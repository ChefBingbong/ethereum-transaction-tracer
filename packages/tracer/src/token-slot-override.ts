import {
  type Address,
  concat,
  encodePacked,
  type Hex,
  hexToBigInt,
  keccak256,
  maxUint128,
  numberToHex,
  pad,
} from 'viem'

function getStandardBalanceSlot(address: Address, slot: bigint): Hex {
  return keccak256(
    encodePacked(['bytes32', 'uint256'], [pad(address, { size: 32 }), slot]),
  )
}

function getStandardAllowanceSlot(
  owner: Address,
  recipient: Address,
  slot: bigint,
): Hex {
  return keccak256(
    encodePacked(
      ['bytes32', 'bytes32'],
      [
        pad(recipient, { size: 32 }),
        keccak256(
          encodePacked(
            ['bytes32', 'uint256'],
            [pad(owner, { size: 32 }), slot],
          ),
        ),
      ],
    ),
  )
}

function getCustomSlotWithAddress(address: Address, baseSlot: Hex): Hex {
  return keccak256(
    encodePacked(
      ['bytes32', 'bytes32'],
      [pad(address, { size: 32 }), baseSlot],
    ),
  )
}

function getSlotForSoladyErc20Approval(owner: Address, recipient: Address) {
  const APPROVE_SLOT_SEED: Hex = '0x7f5e9f20'
  const body = concat([owner, pad(APPROVE_SLOT_SEED, { size: 12 }), recipient])
  return keccak256(body)
}

function getSlotForSoladyErc20Balance(owner: Address) {
  const BALANCE_SLOT_SEED: Hex = '0x87a211a2'
  const body = concat([owner, pad(BALANCE_SLOT_SEED, { size: 12 })])
  return keccak256(body)
}

function getSlotForOZERC20Balance(address: Address) {
  return getStandardBalanceSlot(address, 0n)
}

function getSlotForOZERC20BalanceWithOwner(address: Address) {
  return getStandardBalanceSlot(address, 1n)
}

function getSlotForUSDCBalance(address: Address) {
  return getStandardBalanceSlot(address, 9n)
}

function getSlotForWBTCBalance(address: Address) {
  return getStandardBalanceSlot(address, 5n)
}

function getSlotForUSDT0Balance(address: Address) {
  return getStandardBalanceSlot(address, 51n)
}

function getSlotForWHYPEBalance(address: Address) {
  return getStandardBalanceSlot(address, 3n)
}

function getSlotForOZERC20Allowance(owner: Address, recipient: Address) {
  return getStandardAllowanceSlot(owner, recipient, 1n)
}

function getSlotForOZERC20AllowanceWithOwner(
  owner: Address,
  recipient: Address,
) {
  return getStandardAllowanceSlot(owner, recipient, 2n)
}

function getSlotForUSDCAllowance(owner: Address, recipient: Address) {
  return getStandardAllowanceSlot(owner, recipient, 10n)
}

function getSlotForWBTCAllowance(owner: Address, recipient: Address) {
  return getStandardAllowanceSlot(owner, recipient, 6n)
}

function getSlotForUSDT0Allowance(owner: Address, recipient: Address) {
  return getStandardAllowanceSlot(owner, recipient, 52n)
}

function getSlotForWHYPEAllowance(owner: Address, recipient: Address) {
  return getStandardAllowanceSlot(owner, recipient, 4n)
}

function getSlotForOZUpERC20Balance(address: Address) {
  const baseSlot: Hex =
    '0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00'
  return getCustomSlotWithAddress(address, baseSlot)
}

function getSlotForOZUpERC20Allowance(owner: Address, recipient: Address) {
  const baseSlot =
    hexToBigInt(
      '0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00',
    ) + 1n
  return getStandardAllowanceSlot(owner, recipient, baseSlot)
}

function getSlotForIBGTBalance(address: Address) {
  return getStandardBalanceSlot(address, 2n)
}

function getSlotForIBGTAllowance(owner: Address, recipient: Address) {
  return getStandardAllowanceSlot(owner, recipient, 3n)
}

const maxInt256 = (1n << 255n) - 1n

export function getUnlimitedBalanceAndApprovalStateOverrides(
  owner: Address,
  token: Address,
  approvalRecipient: Address,
  opts: {
    balance?: bigint
    allowance?: bigint
    standardBalanceSlotIndex?: bigint
    standardAllowanceSlotIndex?: bigint
  } = {},
) {
  const {
    balance = maxUint128 - 1n,
    allowance = maxUint128 - 1n,
    standardAllowanceSlotIndex = 4n,
  } = opts

  const balHex = numberToHex(balance, { size: 32 })
  const allHex = numberToHex(allowance, { size: 32 })

  const stateDiff: Record<Hex, Hex> = {
    [getSlotForWHYPEBalance(owner)]: balHex,
    [getStandardAllowanceSlot(
      owner,
      approvalRecipient,
      standardAllowanceSlotIndex,
    )]: allHex,

    [getSlotForWHYPEAllowance(owner, approvalRecipient)]: allHex,
    [getSlotForSoladyErc20Approval(owner, approvalRecipient)]: allHex,
    [getSlotForSoladyErc20Balance(owner)]: balHex,
    [getSlotForOZERC20Allowance(owner, approvalRecipient)]: allHex,
    [getSlotForOZERC20Balance(owner)]: balHex,
    [getSlotForOZERC20AllowanceWithOwner(owner, approvalRecipient)]: allHex,
    [getSlotForOZERC20BalanceWithOwner(owner)]: balHex,
    [getSlotForWBTCBalance(owner)]: balHex,
    [getSlotForWBTCAllowance(owner, approvalRecipient)]: allHex,
    [getSlotForOZUpERC20Allowance(owner, approvalRecipient)]: allHex,
    [getSlotForOZUpERC20Balance(owner)]: balHex,
    [getSlotForUSDCAllowance(owner, approvalRecipient)]: allHex,
    [getSlotForUSDCBalance(owner)]: numberToHex(maxInt256, { size: 32 }),
    [getSlotForUSDT0Balance(owner)]: balHex,
    [getSlotForUSDT0Allowance(owner, approvalRecipient)]: allHex,
    [getSlotForIBGTBalance(owner)]: balHex,
    [getSlotForIBGTAllowance(owner, approvalRecipient)]: allHex,
  }

  return {
    [owner]: {
      balance,
    },
    [token]: {
      stateDiff,
    },
  }
}
