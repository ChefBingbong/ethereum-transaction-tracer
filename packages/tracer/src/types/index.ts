export * from './types.cache'
export * from './types.format'
export * from './types.tracer'
