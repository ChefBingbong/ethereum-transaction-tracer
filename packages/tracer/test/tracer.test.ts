import fs from 'node:fs'
import { afterAll, describe, expect, it } from 'vitest'
import { getUnlimitedBalanceAndApprovalStateOverrides } from '../src'
import {
  defaultTestClient,
  publicTestClient,
  publicTestClient2,
  SAMPLE_TX_HASH,
  SENDER,
  TEST_CACHE_DIR,
  TO,
  TOKEN,
} from './utils/setup'

describe.skip('TransactionTracer â€“ integration (mainnet fork)', () => {
  afterAll(() => {
    try {
      return fs.rmSync(TEST_CACHE_DIR, { recursive: true, force: true })
    } catch {}
  })

  it('formats a real mainnet transaction trace', async () => {
    const [error, res] = await defaultTestClient.traceTransactionHash({
      txHash: SAMPLE_TX_HASH,
      cache: {
        cachePath: TEST_CACHE_DIR,
        etherscanApiKey: '8E6CI28EZUYCY1GG8CMZTPCCCNCVYCS8S2',
      },
      run: {
        env: { kind: 'rpc' },
        showProgressBar: false,
        streamLogs: false,
      },
    })

    expect(error).toBeUndefined()

    expect(res).toBeDefined()
    expect(res?.traceRaw).toBeTruthy()
    expect(res!.traceFormatted).toBeTruthy()
  }, 120_000)

  it('should throw invalid rpc provider error nesecary for trace tx', async () => {
    const [error, res] = await publicTestClient.traceTransactionHash({
      txHash: SAMPLE_TX_HASH,
      cache: {
        cachePath: TEST_CACHE_DIR,
        etherscanApiKey: '8E6CI28EZUYCY1GG8CMZTPCCCNCVYCS8S2',
      },
      run: {
        env: { kind: 'rpc' },
        showProgressBar: false,
        streamLogs: false,
      },
    })

    expect(res).toBeUndefined()
    expect(error?.name).toBeDefined()
    expect(error?.name).toBe('UnsupportedTraceMethodError')
  })

  it('should simulate correctly for trace request', async () => {
    const [error, res] = await publicTestClient2.traceCall({
      account: SENDER,
      blockNumber: 23212888n,
      to: TO,
      data: '0x3593564c000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000068ab678a00000000000000000000000000000000000000000000000000000000000000050a00100604000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000002200000000000000000000000000000000000000000000000000000000000000340000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000008800000000000000000000000000000000000000000000000000000000000000160000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000ffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000068d2ed4e000000000000000000000000000000000000000000000000000000000000000200000000000000000000000066a9893cc07d91d95644aedd05d03f95e1dba8af0000000000000000000000000000000000000000000000000000000068ab675600000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000041c34d9b0f6434f4d6ef00b2b3159d2a1f5988a5537312d06e07c163024a74b330263dfa8956cd830eb6697c45085bcc383702b39fbf1a3ea69ddf8830a6470bcc1c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000015af43ac000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002ba0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000064dac17f958d2ee523a2206206994597c13d831ec700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004a00000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000030b070e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000003800000000000000000000000000000000000000000000000000000000000000060000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec78000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002800000000000000000000000000000000000000000000000000000000000000020000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001f4000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000000000000000000000000000000000271100000000000000000000000000000000000000000000000000000000000000c8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000fee13a103a10d593b9ae06b3e05f2e7e1c00000000000000000000000000000000000000000000000000000000000000190000000000000000000000000000000000000000000000000000000000000060000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000da8a8833e938192781ade161d4b46c4973a40402000000000000000000000000000000000000000000000031f25585d4ebc644aa0c',
      chain: publicTestClient2.chain,
      kzg: undefined,
      stateOverride: getUnlimitedBalanceAndApprovalStateOverrides(
        SENDER,
        TOKEN,
        TO,
      ),
      cache: {
        cachePath: TEST_CACHE_DIR,
        etherscanApiKey: '8E6CI28EZUYCY1GG8CMZTPCCCNCVYCS8S2',
      },
      run: {
        env: { kind: 'fork', blockNumber: 23212888 },
        showProgressBar: false,
        streamLogs: false,
      },
    })

    console.log(res?.traceRaw.calls)
    expect(error).toBeUndefined()
    expect(res).toBeDefined()
    expect(res?.traceFormatted).toMatchSnapshot('traceRaw:summary')
  }, 120000)

  it('should throw invalid rpc provider error nesecary for trace request', async () => {
    const [error, res] = await publicTestClient.traceCall({
      account: SENDER,
      blockNumber: 23212888n,
      to: TO,
      data: '0x3593564c000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000068ab678a00000000000000000000000000000000000000000000000000000000000000050a00100604000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000002200000000000000000000000000000000000000000000000000000000000000340000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000008800000000000000000000000000000000000000000000000000000000000000160000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000ffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000068d2ed4e000000000000000000000000000000000000000000000000000000000000000200000000000000000000000066a9893cc07d91d95644aedd05d03f95e1dba8af0000000000000000000000000000000000000000000000000000000068ab675600000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000041c34d9b0f6434f4d6ef00b2b3159d2a1f5988a5537312d06e07c163024a74b330263dfa8956cd830eb6697c45085bcc383702b39fbf1a3ea69ddf8830a6470bcc1c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000015af43ac000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002ba0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000064dac17f958d2ee523a2206206994597c13d831ec700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004a00000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000030b070e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000003800000000000000000000000000000000000000000000000000000000000000060000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec78000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002800000000000000000000000000000000000000000000000000000000000000020000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001f4000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000000000000000000000000000000000271100000000000000000000000000000000000000000000000000000000000000c8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000fee13a103a10d593b9ae06b3e05f2e7e1c00000000000000000000000000000000000000000000000000000000000000190000000000000000000000000000000000000000000000000000000000000060000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000da8a8833e938192781ade161d4b46c4973a40402000000000000000000000000000000000000000000000031f25585d4ebc644aa0c',
      chain: publicTestClient.chain,
      kzg: undefined,
      stateOverride: getUnlimitedBalanceAndApprovalStateOverrides(
        SENDER,
        TOKEN,
        TO,
      ),
      cache: {
        cachePath: TEST_CACHE_DIR,
        etherscanApiKey: '8E6CI28EZUYCY1GG8CMZTPCCCNCVYCS8S2',
      },
      run: {
        env: { kind: 'rpc' },
        showProgressBar: false,
        streamLogs: false,
      },
    })

    expect(res).toBeUndefined()
    expect(error?.name).toBeDefined()
    expect(error?.name).toBe('UnsupportedTraceMethodError')
  })
})
