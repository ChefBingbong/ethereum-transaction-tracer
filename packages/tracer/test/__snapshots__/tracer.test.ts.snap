// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionTracer ‚Äì integration (mainnet fork) > should throw invalid rpc provider error nesecary for trace request > traceRaw:summary 1`] = `
"UniversalRouter()::execute(commands: 0x0a00100604, inputs: [0x000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000ffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000068d2ed4e000000000000000000000000000000000000000000000000000000000000000200000000000000000000000066a9893cc07d91d95644aedd05d03f95e1dba8af0000000000000000000000000000000000000000000000000000000068ab675600000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000041c34d9b0f6434f4d6ef00b2b3159d2a1f5988a5537312d06e07c163024a74b330263dfa8956cd830eb6697c45085bcc383702b39fbf1a3ea69ddf8830a6470bcc1c00000000000000000000000000000000000000000000000000000000000000, 0x00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000015af43ac000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002ba0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000064dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000, 0x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000030b070e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000003800000000000000000000000000000000000000000000000000000000000000060000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec78000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002800000000000000000000000000000000000000000000000000000000000000020000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001f4000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000000000000000000000000000000000271100000000000000000000000000000000000000000000000000000000000000c8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a9400000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000, 0x000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000fee13a103a10d593b9ae06b3e05f2e7e1c0000000000000000000000000000000000000000000000000000000000000019, 0x000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000da8a8833e938192781ade161d4b46c4973a40402000000000000000000000000000000000000000000000031f25585d4ebc644aa], deadline: 1756063626)  [call] value=0 ETH(gas=0.04496404 used=0.000049471) Gwei ‚ùå
   ‚îî‚îÄ Permit2()::permit(owner: 0xda8A8833E938192781AdE161d4b46c4973A40402, permitSingle: { details: { token: 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48, amount: 1461501637330902918203684832716283019655932542975, expiration: 1758653774, nonce: 2 }, spender: 0x66a9893cC07D91D95644AEDD05D03f95e1dBA8Af, sigDeadline: 1756063574 }, signature: 0xc34d9b0f6434f4d6ef00b2b3159d2a1f5988a5537312d06e07c163024a74b330263dfa8956cd830eb6697c45085bcc383702b39fbf1a3ea69ddf8830a6470bcc1c)  [call] value=0 ETH(gas=0.044256516 used=0.000007928) Gwei ‚ùå
      ‚îî‚îÄ 0x0000000000000000000000000000000000000001::ecrecover recover signer for hash: 0xa738514c236748f0a3f14c125cd859e0‚Ä¶, v: 28, r: 8833812288336084803896656917423426‚Ä¶, s: 1729739505635639503616999239372779‚Ä¶  [staticcall] value=0 ETH(gas=0.043562684 used=0.000003) Gwei
         [Return] signer: 0x000000000000000000000000da8a8833‚Ä¶
      [Revert] InvalidNonce()
   [Revert] ExecutionFailed(0, 0x756688fe)"
`;

exports[`TransactionTracer ‚Äì integration (mainnet fork) should throw invalid rpc provider error nesecary for trace request: traceRaw:summary 1`] = `
"[31m[1mUniversalRouter[2m()[22m[1m[22m[39m::[1m[31mexecute[39m[22m([2mcommands[22m: 0x0a00100604, [2minputs[22m: [0x000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000000000000000000000000ffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000068d2ed4e000000000000000000000000000000000000000000000000000000000000000200000000000000000000000066a9893cc07d91d95644aedd05d03f95e1dba8af0000000000000000000000000000000000000000000000000000000068ab675600000000000000000000000000000000000000000000000000000000000000e00000000000000000000000000000000000000000000000000000000000000041c34d9b0f6434f4d6ef00b2b3159d2a1f5988a5537312d06e07c163024a74b330263dfa8956cd830eb6697c45085bcc383702b39fbf1a3ea69ddf8830a6470bcc1c00000000000000000000000000000000000000000000000000000000000000, 0x00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000015af43ac000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000002ba0b86991c6218b36c1d19d4a2e9eb0ce3606eb48000064dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000, 0x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000030b070e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000003800000000000000000000000000000000000000000000000000000000000000060000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec78000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002800000000000000000000000000000000000000000000000000000000000000020000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001f4000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000000000000000000000000000000000271100000000000000000000000000000000000000000000000000000000000000c8000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a9400000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000, 0x000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000000000fee13a103a10d593b9ae06b3e05f2e7e1c0000000000000000000000000000000000000000000000000000000000000019, 0x000000000000000000000000cccccccccc33d538dbc2ee4feab0a7a1ff4e8a94000000000000000000000000da8a8833e938192781ade161d4b46c4973a40402000000000000000000000000000000000000000000000031f25585d4ebc644aa], [2mdeadline[22m: 1756063626)  [33m[call][39m [2m[2mvalue=[22m[2m0 ETH[22m[2m([2mgas=[22m[2m0.04496404 [2mused=[22m[2m0.000049471) Gwei[22m [31m‚ùå[39m
   ‚îî‚îÄ [31m[1mPermit2[2m()[22m[1m[22m[39m::[1m[31mpermit[39m[22m([2mowner[22m: 0xda8A8833E938192781AdE161d4b46c4973A40402, [2mpermitSingle[22m: { details: { token: 0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48, amount: 1461501637330902918203684832716283019655932542975, expiration: 1758653774, nonce: 2 }, spender: 0x66a9893cC07D91D95644AEDD05D03f95e1dBA8Af, sigDeadline: 1756063574 }, [2msignature[22m: 0xc34d9b0f6434f4d6ef00b2b3159d2a1f5988a5537312d06e07c163024a74b330263dfa8956cd830eb6697c45085bcc383702b39fbf1a3ea69ddf8830a6470bcc1c)  [33m[call][39m [2m[2mvalue=[22m[2m0 ETH[22m[2m([2mgas=[22m[2m0.044256516 [2mused=[22m[2m0.000007928) Gwei[22m [31m‚ùå[39m
      ‚îî‚îÄ [34m[1m0x0000000000000000000000000000000000000001[22m[39m::[1mecrecover[22m recover signer for hash: 0xa738514c236748f0a3f14c125cd859e0‚Ä¶, v: 28, r: 8833812288336084803896656917423426‚Ä¶, s: 1729739505635639503616999239372779‚Ä¶  [33m[staticcall][39m [2m[2mvalue=[22m[2m0 ETH[22m[2m([2mgas=[22m[2m0.043562684 [2mused=[22m[2m0.000003) Gwei[22m
         [32m[Return][39m signer: 0x000000000000000000000000da8a8833‚Ä¶
      [31m[Revert][39m [31mInvalidNonce()[39m
   [31m[Revert][39m [31mExecutionFailed(0, 0x756688fe)[39m"
`;
